File: iEngine-Web/src/textures/TextureUtils.ts
import { TextureWrapMode, TextureMinFilter, TextureMagFilter } from "./Texture";

// WebGL 映射
export function getWebGLWrapMode(gl: WebGLRenderingContext, mode: TextureWrapMode): number {
    switch (mode) {
        case TextureWrapMode.Repeat: return gl.REPEAT;
        case TextureWrapMode.ClampToEdge: return gl.CLAMP_TO_EDGE;
        case TextureWrapMode.MirroredRepeat: return gl.MIRRORED_REPEAT;
        default: return gl.REPEAT;
    }
}

export function getWebGLMinFilter(gl: WebGLRenderingContext, filter: TextureMinFilter): number {
    switch (filter) {
        case TextureMinFilter.Nearest: return gl.NEAREST;
        case TextureMinFilter.Linear: return gl.LINEAR;
        case TextureMinFilter.NearestMipmapNearest: return gl.NEAREST_MIPMAP_NEAREST;
        case TextureMinFilter.LinearMipmapNearest: return gl.LINEAR_MIPMAP_NEAREST;
        case TextureMinFilter.NearestMipmapLinear: return gl.NEAREST_MIPMAP_LINEAR;
        case TextureMinFilter.LinearMipmapLinear: return gl.LINEAR_MIPMAP_LINEAR;
        default: return gl.LINEAR;
    }
}

export function getWebGLMagFilter(gl: WebGLRenderingContext, filter: TextureMagFilter): number {
    switch (filter) {
        case TextureMagFilter.Nearest: return gl.NEAREST;
        case TextureMagFilter.Linear: return gl.LINEAR;
        default: return gl.LINEAR;
    }
}

// WebGPU 映射
export function getWebGPUWrapMode(mode: TextureWrapMode): GPUAddressMode {
    switch (mode) {
        case TextureWrapMode.Repeat: return 'repeat';
        case TextureWrapMode.ClampToEdge: return 'clamp-to-edge';
        case TextureWrapMode.MirroredRepeat: return 'mirror-repeat';
        default: return 'repeat';
    }
}

export function getWebGPUMinMagFilter(filter: TextureMinFilter | TextureMagFilter): GPUFilterMode {
    // WebGPU 只支持 'nearest' | 'linear'
    switch (filter) {
        case TextureMinFilter.Nearest:
        case TextureMagFilter.Nearest:
            return 'nearest';
        case TextureMinFilter.Linear:
        case TextureMagFilter.Linear:
        default:
            return 'linear';
    }
}

File: iEngine-Native/Engine/src/textures/TextureUtils.cpp
#include "iengine/textures/TextureUtils.h"

// OpenGL 头文件 (在实际项目中可能需要包含具体的OpenGL头文件)
// #include <GL/glew.h> 或 #include <glad/glad.h>

namespace iengine {

    unsigned int getOpenGLWrapMode(TextureWrapMode mode) {
        switch (mode) {
            case TextureWrapMode::Repeat:
                // return GL_REPEAT;
                return 0x2901; // GL_REPEAT 的值
            case TextureWrapMode::ClampToEdge:
                // return GL_CLAMP_TO_EDGE;
                return 0x812F; // GL_CLAMP_TO_EDGE 的值
            case TextureWrapMode::MirroredRepeat:
                // return GL_MIRRORED_REPEAT;
                return 0x8370; // GL_MIRRORED_REPEAT 的值
            default:
                // return GL_REPEAT;
                return 0x2901; // GL_REPEAT 的值
        }
    }

    unsigned int getOpenGLMinFilter(TextureMinFilter filter) {
        switch (filter) {
            case TextureMinFilter::Nearest:
                // return GL_NEAREST;
                return 0x2600; // GL_NEAREST 的值
            case TextureMinFilter::Linear:
                // return GL_LINEAR;
                return 0x2601; // GL_LINEAR 的值
            case TextureMinFilter::NearestMipmapNearest:
                // return GL_NEAREST_MIPMAP_NEAREST;
                return 0x2700; // GL_NEAREST_MIPMAP_NEAREST 的值
            case TextureMinFilter::LinearMipmapNearest:
                // return GL_LINEAR_MIPMAP_NEAREST;
                return 0x2701; // GL_LINEAR_MIPMAP_NEAREST 的值
            case TextureMinFilter::NearestMipmapLinear:
                // return GL_NEAREST_MIPMAP_LINEAR;
                return 0x2702; // GL_NEAREST_MIPMAP_LINEAR 的值
            case TextureMinFilter::LinearMipmapLinear:
                // return GL_LINEAR_MIPMAP_LINEAR;
                return 0x2703; // GL_LINEAR_MIPMAP_LINEAR 的值
            default:
                // return GL_LINEAR;
                return 0x2601; // GL_LINEAR 的值
        }
    }

    unsigned int getOpenGLMagFilter(TextureMagFilter filter) {
        switch (filter) {
            case TextureMagFilter::Nearest:
                // return GL_NEAREST;
                return 0x2600; // GL_NEAREST 的值
            case TextureMagFilter::Linear:
                // return GL_LINEAR;
                return 0x2601; // GL_LINEAR 的值
            default:
                // return GL_LINEAR;
                return 0x2601; // GL_LINEAR 的值
        }
    }

} // namespace iengine